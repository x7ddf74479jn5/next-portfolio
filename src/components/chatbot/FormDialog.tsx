import Button from "@material-ui/core/Button";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import React, { useCallback, useState } from "react";
import useClickAway from "src/hooks/useClickAway";
import styles from "src/styles/components/chatbot/FormDialog.module.scss";

import { validateEmailFormat, validateRequiredInput } from "../../utils/helper";
import TextInput from "./TextInput";

type Props = {
  handleClose: () => void;
};

const FormDialog: React.VFC<Props> = (props) => {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [description, setDescription] = useState("");
  const ref = React.useRef<HTMLDivElement>(null);
  useClickAway(ref, props.handleClose);
  const inputName = useCallback(
    (event) => {
      setName(event.target.value);
    },
    [setName]
  );
  const inputEmail = useCallback(
    (event) => {
      setEmail(event.target.value);
    },
    [setEmail]
  );
  const inputDescription = useCallback(
    (event) => {
      setDescription(event.target.value);
    },
    [setDescription]
  );

  const submitForm = () => {
    const isBlank = validateRequiredInput(name, email, description);
    const isValidEmail = validateEmailFormat(email);

    if (isBlank) {
      alert("必須入力欄が空白です。");
      return false;
    } else if (!isValidEmail) {
      alert("メールアドレスの書式が異なります。");
      return false;
    } else {
      const payload = {
        text:
          "お問い合わせがありました\n" +
          "お名前:" +
          name +
          "\n" +
          "Email:" +
          email +
          "\n" +
          "お問い合わせ内容:\n" +
          description,
      };
      const url = process.env.SLACK_API_ENDPOINT;
      if (!url) {
        throw new Error("API endpoint is undefined");
      }

      fetch(url, {
        method: "POST",
        body: JSON.stringify(payload),
      }).then(() => {
        alert("お問い合わせが完了しました。追ってご連絡します");
        setName("");
        setEmail("");
        setDescription("");
        return props.handleClose();
      });
    }
  };

  return (
    <div className={styles.formDialog} ref={ref}>
      <DialogTitle id="alert-dialog-title">お問い合わせフォーム</DialogTitle>
      <DialogContent>
        <TextInput
          label={"件名（必須）"}
          id="name"
          multiline={false}
          rows={1}
          value={name}
          type={"text"}
          onChange={inputName}
        />
        <TextInput
          label={"メールアドレス（必須）"}
          id="email"
          multiline={false}
          rows={1}
          value={email}
          type={"email"}
          onChange={inputEmail}
        />
        <TextInput
          label={"お問い合わせ内容（必須）"}
          id="description"
          multiline={true}
          rows={5}
          value={description}
          type={"text"}
          onChange={inputDescription}
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={props.handleClose} color="primary">
          キャンセル
        </Button>
        <Button onClick={submitForm} color="primary">
          送信する
        </Button>
      </DialogActions>
    </div>
  );
};

export default FormDialog;
