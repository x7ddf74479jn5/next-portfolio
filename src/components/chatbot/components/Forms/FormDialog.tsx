import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import React, { useCallback, useState } from "react";

import TextInput from "./TextInput";

type Props = {
  open: boolean;
  handleClose: () => void;
};

const FormDialog: React.VFC<Props> = (props) => {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [description, setDescription] = useState("");

  const inputName = useCallback(
    (event) => {
      setName(event.target.value);
    },
    [setName]
  );
  const inputEmail = useCallback(
    (event) => {
      setEmail(event.target.value);
    },
    [setEmail]
  );
  const inputDescription = useCallback(
    (event) => {
      setDescription(event.target.value);
    },
    [setDescription]
  );

  const validateEmailFormat = (email: string) => {
    const regex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
    return regex.test(email);
  };

  const validateRequiredInput = (...args: string[]) => {
    let isBlank = false;
    for (let i = 0; i < args.length; i = (i + 1) | 0) {
      if (args[i] === "") {
        isBlank = true;
      }
    }
    return isBlank;
  };

  const submitForm = () => {
    const isBlank = validateRequiredInput(name, email, description);
    const isValidEmail = validateEmailFormat(email);

    if (isBlank) {
      alert("必須入力欄が空白です。");
      return false;
    } else if (!isValidEmail) {
      alert("メールアドレスの書式が異なります。");
      return false;
    } else {
      const payload = {
        text:
          "お問い合わせがありました\n" +
          "お名前:" +
          name +
          "\n" +
          "Email:" +
          email +
          "\n" +
          "お問い合わせ内容:\n" +
          description,
      };
      const url = "https://hooks.slack.com/services/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";

      fetch(url, {
        method: "POST",
        body: JSON.stringify(payload),
      }).then(() => {
        alert("お問い合わせが完了しました。追ってご連絡します");
        setName("");
        setEmail("");
        setDescription("");
        return props.handleClose();
      });
    }
  };

  return (
    <Dialog
      open={props.open}
      onClose={props.handleClose}
      aria-labelledby="alert-dialog-title"
      aria-describedby="alert-dialog-description"
    >
      <DialogTitle id="alert-dialog-title">お問い合わせフォーム</DialogTitle>
      <DialogContent>
        <TextInput label={"件名（必須）"} multiline={false} rows={1} value={name} type={"text"} onChange={inputName} />
        <TextInput
          label={"メールアドレス（必須）"}
          multiline={false}
          rows={1}
          value={email}
          type={"email"}
          onChange={inputEmail}
        />
        <TextInput
          label={"お問い合わせ内容（必須）"}
          multiline={true}
          rows={5}
          value={description}
          type={"text"}
          onChange={inputDescription}
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={props.handleClose} color="primary">
          キャンセル
        </Button>
        <Button onClick={submitForm} color="primary">
          送信する
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default FormDialog;
