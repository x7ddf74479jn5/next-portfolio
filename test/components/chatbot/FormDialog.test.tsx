import userEvent from "@testing-library/user-event";
import FormDialog from "src/components/chatbot/FormDialog";
import type { RenderResult } from "test/test-utils";
import { act, render } from "test/test-utils";

describe("FormDialog", () => {
  let renderResult: RenderResult;
  let nameInput: HTMLElement;
  let emailInput: HTMLElement;
  let descriptionInput: HTMLElement;
  let cancelButton: HTMLElement;
  let submitButton: HTMLElement;
  const mockHandleClose = jest.fn();
  global.fetch = jest.fn().mockResolvedValue("success");
  window.alert = jest.fn();

  beforeEach(() => {
    renderResult = render(<FormDialog handleClose={mockHandleClose} />);
    nameInput = renderResult.getByLabelText(/件名/);
    emailInput = renderResult.getByLabelText(/メールアドレス/);
    descriptionInput = renderResult.getByLabelText(/お問い合わせ内容/);
    cancelButton = renderResult.getAllByRole("button")[0];
    submitButton = renderResult.getAllByRole("button")[1];
    jest.clearAllMocks();
  });

  it("matches snapshot", () => {
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("rendered correctly", () => {
    expect(renderResult.getByRole("heading")).toHaveTextContent("お問い合わせフォーム");
    expect(renderResult.container.firstElementChild).toHaveClass("formDialog");

    expect(nameInput).toHaveAttribute("id", "name");
    expect(nameInput).toHaveAttribute("rows", "1");
    expect(nameInput).toHaveAttribute("type", "text");
    expect(nameInput).toHaveValue("");

    expect(emailInput).toHaveAttribute("id", "email");
    expect(emailInput).toHaveAttribute("rows", "1");
    expect(emailInput).toHaveAttribute("type", "email");
    expect(emailInput).toHaveValue("");

    expect(descriptionInput).toHaveAttribute("id", "description");
    expect(descriptionInput).toHaveAttribute("rows", "5");
    expect(descriptionInput).toHaveValue("");

    expect(cancelButton).toHaveTextContent("キャンセル");
    expect(cancelButton).toBeEnabled();
    expect(submitButton).toHaveTextContent("送信する");
    expect(submitButton).toBeEnabled();
  });

  it("should trigger onChange event", () => {
    userEvent.type(nameInput, "test");
    expect(nameInput).toHaveValue("test");
    userEvent.clear(nameInput);
    expect(nameInput).not.toHaveValue("test");
  });

  it("should success to submit with valid values", async () => {
    userEvent.type(nameInput, "name");
    userEvent.type(emailInput, "email@test.com");
    userEvent.type(descriptionInput, "description");
    expect(nameInput).toHaveValue("name");
    expect(emailInput).toHaveValue("email@test.com");
    expect(descriptionInput).toHaveValue("description");

    await act(async () => {
      await userEvent.click(submitButton);
    });
    const url = process.env.SLACK_API_ENDPOINT;
    const option = expect.objectContaining({
      method: "POST",
      body: expect.stringMatching(/.*name.*email@test\.com.*description.*/),
    });
    expect(global.fetch).toBeCalledWith(url, option);
    expect(window.alert).toBeCalledWith(expect.stringContaining("完了"));
    expect(mockHandleClose).toBeCalled();
  });

  it("should fail to submit with missing required values", async () => {
    expect(nameInput).toHaveValue("");
    expect(emailInput).toHaveValue("");
    expect(descriptionInput).toHaveValue("");

    await act(async () => {
      await userEvent.click(submitButton);
    });
    expect(global.fetch).not.toBeCalled();
    expect(window.alert).toBeCalledWith(expect.stringContaining("必須"));
  });

  it("should fail to submit with an invalid email value", async () => {
    userEvent.type(nameInput, "name");
    userEvent.type(emailInput, "email");
    userEvent.type(descriptionInput, "description");
    expect(nameInput).toHaveValue("name");
    expect(emailInput).toHaveValue("email");
    expect(descriptionInput).toHaveValue("description");

    await act(async () => {
      await userEvent.click(submitButton);
    });

    expect(global.fetch).not.toBeCalled();
    expect(window.alert).toBeCalledWith(expect.stringContaining("メールアドレス"));
  });

  it("should cancel and close the dialog", () => {
    userEvent.click(cancelButton);
    expect(mockHandleClose).toBeCalled();
  });
});
