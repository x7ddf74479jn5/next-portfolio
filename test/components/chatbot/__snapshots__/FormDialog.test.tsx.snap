// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDialog matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="formDialog"
  >
    <div
      class="MuiDialogTitle-root"
      id="alert-dialog-title"
    >
      <h2
        class="MuiTypography-root MuiTypography-h6"
      >
        お問い合わせフォーム
      </h2>
    </div>
    <div
      class="MuiDialogContent-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
          data-shrink="false"
          for="name"
          id="name-label"
        >
          件名（必須）
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            id="name"
            rows="1"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          メールアドレス（必須）
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            id="email"
            rows="1"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
          data-shrink="false"
          for="description"
          id="description-label"
        >
          お問い合わせ内容（必須）
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline MuiInputBase-marginDense MuiInput-marginDense"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            id="description"
            rows="5"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          キャンセル
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          送信する
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
